envoy/http/async_client.h

```diff
@@ -444,6 +449,8 @@ class AsyncClient {
     absl::optional<bool> sampled_{true};
     // The pointer to sidestream watermark callbacks. Optional, nullptr by default.
     Http::SidestreamWatermarkCallbacks* sidestream_watermark_callbacks = nullptr;
+    // MetadataMatchCriteria to be associated to the new stream
+    const Envoy::Router::MetadataMatchCriteria* metadata_match;
```

Comment:
Now we're carrying metadata information here. It's a pointer, and in my understanding the lifetime of this object is attached to the filter's, and it's hard for me to grasp how safe that is. Any reasons why not using `shared_ptr` in cases like this?

Reply:
Something else that crosses my mind is if this really belongs here. If we have other users of `StreamOptions` that may not have a well defined meaning for a `MetadataMatchCriteria`.

Reply:
To not worry about the lifetime, I think, it might make sense to just keep here a pointer to the original matched route, rather than extract MetadataMatchCriteria out of it. The benefit is that the original route is a shared pointer (see https://github.com/envoyproxy/envoy/blob/main/source/common/router/router.h#L573) and you don't need to worry about lifetimes this way.

The downside is that Route provides access to bunch of metadata that is not really relevant in the context where it will be used, so it will expose more info than strictly necessary.

I don't think that we are losing anything in terms of memory or performance, because the original route will have to be kept around as long as you keep the pointer to MetadataMatchCriteria anyways.

Reply:
I'm not following, what makes RouteEntry to carry a reference to the Route? When you have Route you have access to all the details of Route entries including matching metadata, why do you need to create additional references to anything?